platforms/core-configuration/model-core/src/main/java/org/gradle/api/internal/provider/AbstractProperty.java

```diff
@@ -178,6 +179,20 @@ private Value<? extends T> doCalculateValue(EvaluationScopeContext context, Valu
         }
     }
 
+    /**
+     * Throws a {@link ConcurrentModificationException} with text describing data corruption because of unsafe property access.
+     *
+     * @param reason the (optional) reason indicating the detected corruption
+     * @return nothing (it throws), but you can use this method in {@code throw} statement to appease the compiler
+     */
+    protected ConcurrentModificationException failWithCorruptedStateException(@Nullable Throwable reason) {
+        throw new ConcurrentModificationException(
```

Comment:
🤔 I wonder if it wouldn't be better to return the exception from here (instead of throwing it) so its stack trace is only filled starting with the caller...

Reply:
> 🤔 I wonder if it wouldn't be better to return the exception from here (instead of throwing it) so its stack trace is only filled starting with the caller...

Stack traces are initialized in Exception's constructor, so it wouldn't help, unless I do some tricks.

Reply:
Ah, I guess I was thinking dotnet...